### PR TITLE
Enhancement/Domain command improvements for programming-exercises

### PR DESCRIPTION
**Pull request by thilo-behnke**
### Checklist
- [x] I run `yarn run webpack:build:main`: the project builds without errors.
- [x] I run `yarn lint`: the project builds without code style warnings.
- [x] I updated the documentation and models.
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] ~I added (end-to-end) test cases for the new functionality.~

### Motivation and Context

Some improvements for creating interactive instructions for programming exercises.
Should make working on the instructions more efficient :)

### Description

1. Sort test cases in dropdown alphabetically
2. Make insert test case command more intelligent:
- Make sure there are no duplicates in the brackets
- Add test cases to to beginning or end of list when the cursor is at the beginning or end
- Replace existing test case when cursor is on test case
3. Provide auto completion for test cases
4. Provide auto completion for inserting tasks
5. Turn tasks into an ordered list
6. Provide test case analysis:
- invalid testCases in markdown
- missing testCases in markdown

### Steps for Testing

1. Log in to ArTEMiS
2. Navigate to Course Administration
3. ...

### Screenshots

**1. Alphabetical Sort:**

image

**3. Autocomplete for tests:**

image

**4. Autocomplete for task:**
image

**6. Test Case Analysis:**

image

### CODE REVIEW

File: src/main/webapp/app/markdown-editor/markdown-editor.component.ts [hunk 1]
```diff
@@ -145,6 +146,13 @@ export class MarkdownEditorComponent implements AfterViewInit {
     }
 
     ngAfterViewInit(): void {
+        // Commands may want to add custom completers - remove standard completers of the ace editor.
+        this.aceEditorContainer.getEditor().setOptions({
+            enableBasicAutocompletion: true,
+            enableLiveAutocompletion: true,
+        });
+        this.aceEditorContainer.getEditor().completers = [];
```
Comment (thilo-behnke) [inline src/main/webapp/app/markdown-editor/markdown-editor.component.ts]: This removes the basic (local) auto completion from ace that provides dictionary completion (e.g. be -> between). I remove them here as they don't really help us (and nobody is using them atm anyway).

File: src/main/webapp/app/markdown-editor/markdown-editor.component.ts [hunk 2]
```diff
@@ -239,7 +247,10 @@ export class MarkdownEditorComponent implements AfterViewInit {
             let remainingMarkdownText = this.markdown.slice(0);
 
             /** create string with the identifiers to use for RegEx by deleting the [] of the domainCommandIdentifiers */
-            const commandIdentifiersString = domainCommandIdentifiersToParse.map(tag => tag.replace('[', '').replace(']', '')).join('|');
+            const commandIdentifiersString = domainCommandIdentifiersToParse
+                .map(tag => tag.replace('[', '').replace(']', ''))
+                .map(escapeStringForUseInRegex)
```
Comment (thilo-behnke) [inline src/main/webapp/app/markdown-editor/markdown-editor.component.ts]: If we don't do the escape here, domain commands that use special characters lead to an error when creating the Regex down below.
I had this issue for the "(" character of the testCommand.

**Comment by krusche** [comment]
Very nice work. Can we integrate this into the bugfix release 3.1.5?
I plan to release it tomorrow (Monday)

**Comment by thilo-behnke** [comment]
I want to test it on the testserver this morning, then it should be good to go.

File: src/main/webapp/app/entities/programming-exercise/instructions/programming-exercise-instruction-testcase-status.component.html [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<span *ngIf="invalidTestCases.length + missingTestCases.length > 0">
+    <fa-icon icon="exclamation-triangle" size="2x" class="text-warning" [ngbTooltip]="tooltipContent"></fa-icon>
+</span>
+
+<ng-template #tooltipContent>
+    <div class="text-left w-100">
+        <div *ngIf="invalidTestCases.length">
+            <span>There are {{ invalidTestCases.length }} test cases that are part of the problem statement but can't be found in your test repository:</span>
```
Comment (thilo-behnke) [inline src/main/webapp/app/entities/programming-exercise/instructions/programming-exercise-instruction-testcase-status.component.html]: This needs to be removed into the translation files.

**Comment by thilo-behnke** [comment]
@krusche ready for review :)

**Pull request merged**